[BUG] Windows symlink handling

**Description**
Debugging the test failures for https://github.com/saltstack/salt/pull/67073 on Windows, I discovered three issues related to Windows symlink handling.

Note that I don't have a Windows minion to give proper replication instructions.

1) Symlinks with relative targets pointing to directories create somewhat broken symlinks. This is caused by Windows (file systems?) discriminating between symlinks pointing to files and those that point to directories. The corresponding handling expects symlink targets to be specified in an absolute manner (or relative to the process' CWD, which is unintended for sure) and to exist:

https://github.com/saltstack/salt/blob/4c1b4236b329deea91cd53b33b4a1b0291e72b65/salt/modules/win_file.py#L1464

Rough reproduction:

```sls
{%- set tempdir = salt["temp.dir"]() %}

target_dir:
  file.directory:
    - name: {{ tempdir | path_join("target_dir") }}

symlink:
  file.symlink:
    - name: {{ tempdir | path_join("symlink") }}
    - target: target_dir
    - require:
      - target_dir
```

Symlinks created in this way raised PermissionErrors during specific assertions that called `.stat()` on the associated `pathlib.Path`, such as `.exists()` and `.is_dir()`:

Test runner output:
```console
[...]
assert (base / "symlink_target_dir_overridden_by_dir").is_dir()
[...]
artifacts\salt\lib\pathlib.py:1305: in is_dir
    return S_ISDIR(self.stat().st_mode)
        self       = WindowsPath('D:/a/_temp/pytest-of-runneradmin/pytest-1/test_recurse_merge_True_True_0/target/symlink_target_dir_overridden_by_dir')
[...]
self = WindowsPath('D:/a/_temp/pytest-of-runneradmin/pytest-1/test_recurse_merge_True_True_0/target/symlink_target_dir_overridden_by_dir')

    def stat(self, *, follow_symlinks=True):
        """
        Return the result of the stat() system call on this path, like
        os.stat() does.
        """
>       return self._accessor.stat(self, follow_symlinks=follow_symlinks)
E       PermissionError: [WinError 5] Access is denied: 'D:\\a\\_temp\\pytest-of-runneradmin\\pytest-1\\test_recurse_merge_True_True_0\\target\\symlink_target_dir_overridden_by_dir'
```

Paths below this broken symlink still work correctly though.

2) Similar behavior as in 1) was likely observed with `file.recurse` and `keep_symlinks=True` since the symlinks were created before the directories they pointed to (on top of relative symlinks always being broken).

3) The `roots` fileserver module fails to transform symlink target paths to POSIX style paths when listing symlinks on Windows, which makes its output depend on the master's OS and thus symlink target handling unreliable.

(Not completely sure how to do this on Windows via CLI, sorry!)

```sls
{%- set root_dir = salt["config.get"]("file_roots")["base"] | first %}
{%- set test_dir = root_dir | path_join("testdir") %}

target_file:
  file.managed:
    - name: {{ test_dir | path_join("subdir", "file") }}
    - contents: 'hi'
    - makedirs: true

symlink:
  file.symlink:
    - name: {{ test_dir | path_join("symlink") }}
    - target: subdir/file
    - require:
      - target_file
```

```bash
$ salt-run fileserver.clear_file_list_cache
$ salt-run fileserver.update
$ salt-call cp.list_master_symlinks
local:
    ----------
    testdir/symlink:
        subdir\file
```

**Setup**
This came up in CI (based on current master branch) for Windows 2022 Server runners.

**Steps to Reproduce the behavior**
See above.

**Expected behavior**
1+2) Produce a directory symlink that works.
3) Always return POSIX style paths, otherwise the output cannot be used programmatically.

**Screenshots**
See above.

**Versions Report**
current master branch (commit 1cd6139479)

--- a/changelog/67072.added.md
+++ b/changelog/67072.added.md
@@ -1,0 +1,1 @@
+Added a `merge` option to `file.recurse`, which merges subpaths from all existing `source`s before managing the directory. Handy when using different saltenvs or the TOFS pattern.

--- a/changelog/67766.fixed.md
+++ b/changelog/67766.fixed.md
@@ -1,0 +1,1 @@
+Fixed creating relative directory symlinks on Windows, ensured listing targets of symlinks in file_roots always produces POSIX-style paths

--- a/salt/fileserver/roots.py
+++ b/salt/fileserver/roots.py
@@ -413,6 +413,12 @@ def _file_lists(load, form):
                             abs_path,
                         )
                         link_dest = abs_path
+                    # Not sure what the purpose of this is since symlinks that point outside
+                    # the file roots are allowed (when following symlinks). Either way, this does not do what
+                    # it's intended to do since a symlink that starts with ../ is not resolved
+                    # relative to its full path, but to the containing directory as well.
+                    # This allows symlinks to point to the parent and sibling directories of the file root
+                    # and still be listed here.
                     if link_dest.startswith(".."):
                         joined = os.path.join(abs_path, link_dest)
                     else:
@@ -428,10 +434,10 @@ def _file_lists(load, form):
                         # Only count the link if it does not point
                         # outside of the root dir of the fileserver
                         # (i.e. the "path" variable)
-                        ret["links"][rel_path] = link_dest
+                        ret["links"][rel_path] = _translate_sep(link_dest)
                     else:
                         if not __opts__["fileserver_followsymlinks"]:
-                            ret["links"][rel_path] = link_dest
+                            ret["links"][rel_path] = _translate_sep(link_dest)
 
         for path in __opts__["file_roots"][saltenv]:
             if saltenv == "__env__":

--- a/salt/modules/win_file.py
+++ b/salt/modules/win_file.py
@@ -1462,7 +1462,7 @@ def symlink(src, link, force=False, atomic=False, follow_symlinks=True):
     src = os.path.normpath(src)
     link = os.path.normpath(link)
 
-    is_dir = os.path.isdir(src)
+    is_dir = os.path.isdir(os.path.join(os.path.dirname(link), src))
 
     # Elevate the token from the current process
     desired_access = win32security.TOKEN_QUERY | win32security.TOKEN_ADJUST_PRIVILEGES

--- a/salt/states/file.py
+++ b/salt/states/file.py
@@ -282,7 +282,6 @@ import difflib
 import itertools
 import logging
 import os
-import pathlib
 import posixpath
 import re
 import shutil
@@ -295,6 +294,7 @@ from collections import defaultdict
 from collections.abc import Iterable, Mapping
 from datetime import date, datetime  # python3 problem in the making?
 from itertools import zip_longest
+from pathlib import Path
 
 import salt.loader
 import salt.payload
@@ -431,153 +431,270 @@ def _salt_to_os_path(path):
     return os.path.normpath(path.replace(posixpath.sep, os.path.sep))
 
 
+def _filter_recurse_sources(sources):
+    """
+    Adaptation of ``file.source_list`` for file.recurse. Returns a list
+    of all ``salt://`` scheme sources that represent exisiting directories.
+    """
+    source_list = _validate_str_list(sources)
+
+    for idx, val in enumerate(source_list):
+        source_list[idx] = val.rstrip("/")
+
+    for precheck in source_list:
+        if not precheck.startswith("salt://"):
+            raise CommandExecutionError(
+                f"Invalid source '{precheck}' (must be a salt:// URI)"
+            )
+
+    contextkey = f"recurse__|-{source_list}__|-{__env__}"
+    if contextkey in __context__:
+        return __context__[contextkey]
+    mdirs = {}
+    ret = []
+    for single in source_list:
+        path, senv = salt.utils.url.parse(single)
+        if not senv:
+            senv = __env__
+        if senv not in mdirs:
+            mdirs[senv] = __salt__["cp.list_master_dirs"](saltenv=senv)
+        if path in mdirs[senv]:
+            ret.append(single)
+    if not ret:
+        raise CommandExecutionError("none of the specified sources were found")
+    __context__[contextkey] = ret
+    return __context__[contextkey]
+
+
 def _gen_recurse_managed_files(
     name,
-    source,
+    sources,
     keep_symlinks=False,
     include_pat=None,
     exclude_pat=None,
     maxdepth=None,
     include_empty=False,
+    merge=False,
     **kwargs,
 ):
     """
     Generate the list of files managed by a recurse state
     """
+    # absolute filesystem path of regular file => salt:// source URI with saltenv
+    managed_files = {}
+    # absolute paths of directories that are direct parents of files + empty ones if include_empty
+    managed_directories = set()
+    # absolute filesystem path of symlink => verbatim symlink target spec (relative or absolute, POSIX pathsep)
+    managed_symlinks = {}
+    # set of known managed filesystem paths that should not be deleted when `clean` is True
+    keep = set()
+    # all direct and indirect parent directories involved in this operation, needed for merging sources
+    all_directories = set()
+    ignored_directories = set()
+
+    if not merge:
+        try:
+            sources = [sources[0]]
+        except KeyError:
+            sources = []
 
     # Convert a relative path generated from salt master paths to an OS path
     # using "name" as the base directory
     def full_path(master_relpath):
         return os.path.join(name, _salt_to_os_path(master_relpath))
 
-    # Process symlinks and return the updated filenames list
-    def process_symlinks(filenames, symlinks):
-        for lname, ltarget in symlinks.items():
-            srelpath = posixpath.relpath(lname, srcpath)
-            if not _is_valid_relpath(srelpath, maxdepth=maxdepth):
-                continue
-            if not salt.utils.stringutils.check_include_exclude(
-                srelpath, include_pat, exclude_pat
-            ):
-                continue
-            # Check for all paths that begin with the symlink
-            # and axe it leaving only the dirs/files below it.
-            # This needs to use list() otherwise they reference
-            # the same list.
-            _filenames = list(filenames)
-            for filename in _filenames:
-                if filename.startswith(lname + os.sep):
-                    log.debug(
-                        "** skipping file ** %s, it intersects a symlink", filename
-                    )
-                    filenames.remove(filename)
-            # Create the symlink along with the necessary dirs.
-            # The dir perms/ownership will be adjusted later
-            # if needed
-            managed_symlinks.add((srelpath, ltarget))
-
-            # Add the path to the keep set in case clean is set to True
-            keep.add(full_path(srelpath))
-        vdir.update(keep)
-        return filenames
-
-    managed_files = set()
-    managed_directories = set()
-    managed_symlinks = set()
-    keep = set()
-    vdir = set()
-
-    srcpath, senv = salt.utils.url.parse(source)
-    if senv is None:
-        senv = __env__
-    if not srcpath.endswith(posixpath.sep):
-        # we're searching for things that start with this *directory*.
-        srcpath = srcpath + posixpath.sep
-    fns_ = __salt__["cp.list_master"](senv, srcpath)
-
-    # If we are instructed to keep symlinks, then process them.
-    if keep_symlinks:
-        # Make this global so that emptydirs can use it if needed.
-        symlinks = __salt__["cp.list_master_symlinks"](senv, srcpath)
-        fns_ = process_symlinks(fns_, symlinks)
-
-    for fn_ in fns_:
-        if not fn_.strip():
-            continue
-
-        # fn_ here is the absolute (from file_roots) source path of
-        # the file to copy from; it is either a normal file or an
-        # empty dir(if include_empty==true).
-
-        relname = salt.utils.data.decode(posixpath.relpath(fn_, srcpath))
-        if not _is_valid_relpath(relname, maxdepth=maxdepth):
-            continue
+    # Check whether we want to consider a path at all.
+    # If so, cut the recurse root from the passed path.
+    def check_relative_path(fileserver_path, fs_recurse_base_dir):
+        if not fileserver_path.strip():
+            return False
+        # Render the file path relative to the source (== target) dir
+        relative_path = salt.utils.data.decode(
+            posixpath.relpath(fileserver_path, fs_recurse_base_dir)
+        )
+        if not _is_valid_relpath(relative_path, maxdepth=maxdepth):
+            return False
 
         # Check if it is to be excluded. Match only part of the path
-        # relative to the target directory
+        # relative to the target directory.
         if not salt.utils.stringutils.check_include_exclude(
-            relname, include_pat, exclude_pat
+            relative_path, include_pat, exclude_pat
         ):
-            continue
-        dest = full_path(relname)
-        dirname = os.path.dirname(dest)
-        keep.add(dest)
+            return False
+        return relative_path
 
-        if dirname not in vdir:
-            # verify the directory perms if they are set
-            managed_directories.add(dirname)
-            vdir.add(dirname)
+    # Check whether an absolute destination path is already occupied
+    # for the current operation (only relevant when merging multiple sources).
+    def is_occupied(dest):
+        dpath = Path(dest)
+        return str(dest) in all_directories or any(
+            path in aggr
+            for path in [str(dest), *(str(parent) for parent in dpath.parents)]
+            for aggr in (managed_files, managed_symlinks)
+        )
 
-        src = salt.utils.url.create(fn_, saltenv=senv)
-        managed_files.add((dest, src))
-
-    if include_empty:
-        mdirs = __salt__["cp.list_master_dirs"](senv, srcpath)
-        for mdir in mdirs:
-            relname = posixpath.relpath(mdir, srcpath)
-            if not _is_valid_relpath(relname, maxdepth=maxdepth):
-                continue
-            if not salt.utils.stringutils.check_include_exclude(
-                relname, include_pat, exclude_pat
-            ):
-                continue
-            mdest = full_path(relname)
-            # Check for symlinks that happen to point to an empty dir.
-            if keep_symlinks:
-                islink = False
-                for link in symlinks:
-                    if mdir.startswith(link + os.sep, 0):
+    # When keep_symlinks is true, check which listed fileserver paths are symlink
+    # artifacts and remove them from the list of files to manage.
+    # Add them to the symlink collection if appropriate.
+    def process_symlinks(filenames):
+        # Returns a dict where keys are paths relative to the fileserver root
+        # and values the symlink targets (exactly as specified, relative or absolute).
+        symlinks = __salt__["cp.list_master_symlinks"](senv, recurse_root)
+        for lname, ltarget in symlinks.items():
+            # The fileclient (usually) follows symlinks when listing paths.
+            # We need to remove the symlink and (if it points to a directory)
+            # paths below it from the regular files list, which is passed to file.managed.
+            # If we don't remove the symlink itself, it causes https://github.com/saltstack/salt/issues/64630
+            # when the target file is not created before the symlink is processed by file.managed
+            #   (specifically in file.manage_file, which tries to hash the symlink target when
+            #    managing the symlink and follow_symlinks was not set to False).
+            # This issue was introduced by https://github.com/saltstack/salt/commit/fd7e82fe678906faff2610575164fb799ff489ef
+            # The referenced commit sought to fix another issue, which caused
+            # all paths that had the same prefix as the symlink to be discarded.
+            # As a side effect, it unnecessarily subjected all symlinks to file.managed processing.
+            # This bug was worked around in https://github.com/saltstack/salt/pull/66857 by
+            # ensuring symlinks were always managed last, but they don't need to be
+            # file.managed in the first place.
+            target_was_dir = False
+            for filename in list(filenames):
+                if filename == lname or filename.startswith(lname + posixpath.sep):
+                    filenames.remove(filename)
+                    if filename != lname:
                         log.debug(
-                            "** skipping empty dir ** %s, it intersects a symlink", mdir
+                            "** skipping file ** %s, it intersects a symlink", filename
                         )
-                        islink = True
-                        break
-                if islink:
-                    continue
+                        target_was_dir = True
 
-            managed_directories.add(mdest)
-            keep.add(mdest)
+            # This check needs to happen after removing all artifacts of the symlink
+            # from the list of managed files, otherwise the symlink itself would be ignored,
+            # but the data it pointed to would be managed as regular files/directories.
+            # If accepted, this contains the path of the symlink relative to the recurse root
+            # (still using POSIX pathsep)
+            relative_path = check_relative_path(lname, recurse_root)
+            if relative_path is False:
+                continue
+            # This is the path on the minion where the symlink should be placed (using OS-specific pathsep)
+            dest = full_path(relative_path)
+            if is_occupied(dest):
+                continue
 
-    # Sets are randomly ordered. We need to use a list so we can make sure
-    # symlinks are always at the end. This is necessary because the file must
-    # exist before we can create a symlink to it. See issue:
-    # https://github.com/saltstack/salt/issues/64630
-    new_managed_files = list(managed_files)
-    # Now let's move all the symlinks to the end
-    for link_src_relpath, _ in managed_symlinks:
-        for file_dest, file_src in managed_files:
-            # We need to convert relpath to fullpath. We're using pathlib to
-            # be platform-agnostic
-            symlink_full_path = pathlib.Path(f"{name}{os.sep}{link_src_relpath}")
-            file_dest_full_path = pathlib.Path(file_dest)
-            if symlink_full_path == file_dest_full_path:
-                new_managed_files.append(
-                    new_managed_files.pop(
-                        new_managed_files.index((file_dest, file_src))
-                    )
+            # This is the path the symlink points to on the minion (using OS-specific pathsep)
+            local_symlink_target = os.path.normpath(
+                os.path.join(
+                    os.path.dirname(dest), ltarget.replace(posixpath.sep, os.path.sep)
                 )
+            )
+            # Ensure symlink targets don't change type because of merging, otherwise drop the symlink.
+            # This can only happen for symlinks within the recurse root.
+            if target_was_dir and local_symlink_target in managed_files:
+                # We need to ignore it explicitly (when include_empty is true),
+                # otherwise the directory would be detected as an empty one.
+                ignored_directories.add(dest)
+                continue
+            if not target_was_dir and local_symlink_target in managed_directories:
+                continue
+            # If we're here, the symlink target is not managed as part of this recurse,
+            # is the same as the symlink pointed to originally or is of the same type.
+            # Remember the link target as returned by the fileserver (POSIX pathsep) because
+            # we might need it for filtering empty directories later.
+            managed_symlinks[dest] = ltarget
 
-    return new_managed_files, managed_directories, managed_symlinks, keep
+            # Remember the symlink path as being managed so it's not deleted when clean=True
+            keep.add(dest)
+        return filenames
+
+    for source in sources:
+        recurse_root, senv = salt.utils.url.parse(source)
+        if senv is None:
+            senv = __env__
+        if not recurse_root.endswith(posixpath.sep):
+            # we're searching for things that start with this *directory*.
+            recurse_root += posixpath.sep
+
+        # Fetch a list of all files in the current source.
+        # Paths are relative to the fileserver root and use the POSIX path separator (/).
+        # Includes resolved symlinks (i.e. directory symlinks are resolved to the contents of the target dir).
+        fns_ = __salt__["cp.list_master"](senv, recurse_root)
+
+        # If we are instructed to keep symlinks, we need to reverse their effects on the file list.
+        if keep_symlinks:
+            fns_ = process_symlinks(fns_)
+
+        for fn_ in fns_:
+            relative_path = check_relative_path(fn_, recurse_root)
+            if relative_path is False:
+                continue
+
+            dest = full_path(relative_path)
+            dpath = Path(dest)
+            if is_occupied(dpath):
+                # We're merging multiple sources and a higher-priority one
+                # provides the same path or a parent one that's not a directory.
+                continue
+            keep.add(dest)
+
+            managed_directories.add(str(dpath.parent))
+            all_directories.update(str(par) for par in dpath.parents)
+
+            salt_uri_with_senv = salt.utils.url.create(fn_, saltenv=senv)
+            managed_files[dest] = salt_uri_with_senv
+
+        # Empty directories are not contained in the list of files from above.
+        if include_empty:
+            mdirs = __salt__["cp.list_master_dirs"](senv, recurse_root)
+            for mdir in mdirs:
+                relative_path = check_relative_path(mdir, recurse_root)
+                if relative_path is False:
+                    continue
+                dest = full_path(relative_path)
+                if dest in ignored_directories or is_occupied(dest):
+                    # In addition to the usual checks, don't try to create
+                    # directories that served as symlink targets in the current
+                    # source layer and were overridden with a file in a higher
+                    # priority one.
+                    continue
+                managed_directories.add(dest)
+                all_directories.add(dest)
+                keep.add(dest)
+
+        elif keep_symlinks:
+            # Filter invalid symlinks caused by pointing to an empty directory
+            # and include_empty being false. This needs to happen recursively
+            # because symlinks can point to other symlinks.
+            removed_links = None
+            mdirs = __salt__["cp.list_master_dirs"](senv, recurse_root)
+            while removed_links is None or removed_links:
+                removed_links = 0
+                for lnk in list(managed_symlinks):
+                    relative_target = managed_symlinks[lnk]
+                    if (
+                        posixpath.normpath(
+                            posixpath.join(recurse_root, relative_target)
+                        )
+                        not in mdirs
+                    ):
+                        # The symlink never pointed to a recursed subdirectory in the first place
+                        continue
+                    full_tgt = os.path.normpath(full_path(relative_target))
+                    if any(
+                        full_tgt in aggr
+                        for aggr in (managed_files, all_directories, managed_symlinks)
+                    ):
+                        # The target exists in some way
+                        continue
+                    managed_symlinks.pop(lnk)
+                    removed_links += 1
+
+    # We need to ensure the link target spec respects the host pathsep (it's still POSIX at this point)
+    return (
+        set(managed_files.items()),
+        managed_directories,
+        {
+            (link_path, link_target.replace(posixpath.sep, os.path.sep))
+            for link_path, link_target in managed_symlinks.items()
+        },
+        keep,
+    )
 
 
 def _gen_keep_files(name, require, walk_d=None):
@@ -643,9 +760,21 @@ def _gen_keep_files(name, require, walk_d=None):
                     if os.path.isdir(fn):
                         if _is_child(fn, name):
                             if fun == "recurse":
-                                fkeep = _gen_recurse_managed_files(**low)[3]
-                                log.debug("Keep from %s: %s", fn, fkeep)
-                                keep.update(fkeep)
+                                try:
+                                    sources = _filter_recurse_sources(
+                                        low.get("source", [])
+                                    )
+                                except CommandExecutionError:
+                                    pass
+                                else:
+                                    recurse_kwargs = low.copy()
+                                    recurse_kwargs.pop("source", None)
+                                    recurse_kwargs["sources"] = sources
+                                    fkeep = _gen_recurse_managed_files(
+                                        **recurse_kwargs
+                                    )[3]
+                                    log.debug("Keep from %s: %s", fn, fkeep)
+                                    keep.update(fkeep)
                             elif walk_d:
                                 walk_ret = set()
                                 _process_by_walk_d(fn, walk_ret)
@@ -4496,6 +4625,7 @@ def recurse(
     win_perms=None,
     win_deny_perms=None,
     win_inheritance=True,
+    merge=False,
     **kwargs,
 ):
     """
@@ -4710,6 +4840,23 @@ def recurse(
 
         .. versionadded:: 2017.7.7
 
+    merge
+        When ``source`` is a list, instead of choosing the first existing
+        source as the single source directory, merge paths in all existing
+        sources. When the same path exists in several sources, the earliest
+        source has priority. Defaults to false.
+
+        .. versionadded:: 3008.0
+
+        .. note::
+            When ``keep_symlinks`` is true, symlinks are treated like files,
+            i.e. override lower priority files, directories and symlinks completely.
+            When a higher priority source overrides the symlink target with the
+            same type (file -> file/dir -> dir), they are kept as expected.
+            Special handling is given to symlinks whose target type changes
+            because of the override (file -> dir/dir -> file), where the
+            (lower priority) symlink is dropped, even if it is not overridden
+            by a higher priority source.
     """
     if "env" in kwargs:
         # "env" is not supported; Use "saltenv".
@@ -4767,39 +4914,12 @@ def recurse(
     if not os.path.isabs(name):
         return _error(ret, f"Specified file {name} is not an absolute path")
 
-    # expand source into source_list
-    source_list = _validate_str_list(source)
-
-    for idx, val in enumerate(source_list):
-        source_list[idx] = val.rstrip("/")
-
-    for precheck in source_list:
-        if not precheck.startswith("salt://"):
-            return _error(
-                ret,
-                f"Invalid source '{precheck}' (must be a salt:// URI)",
-            )
-
-    # Select the first source in source_list that exists
+    # Validate the sources and ensure that at least one exists.
     try:
-        source, source_hash = __salt__["file.source_list"](source_list, "", __env__)
+        sources = _filter_recurse_sources(source)
     except CommandExecutionError as exc:
         ret["result"] = False
         ret["comment"] = f"Recurse failed: {exc}"
-        return ret
-
-    # Check source path relative to fileserver root, make sure it is a
-    # directory
-    srcpath, senv = salt.utils.url.parse(source)
-    if senv is None:
-        senv = __env__
-    master_dirs = __salt__["cp.list_master_dirs"](saltenv=senv, prefix=srcpath + "/")
-    if srcpath not in master_dirs:
-        ret["result"] = False
-        ret["comment"] = (
-            "The directory '{}' does not exist on the salt fileserver "
-            "in saltenv '{}'".format(srcpath, senv)
-        )
         return ret
 
     # Verify the target directory
@@ -4908,12 +5028,25 @@ def recurse(
         merge_ret(path, _ret)
 
     mng_files, mng_dirs, mng_symlinks, keep = _gen_recurse_managed_files(
-        name, source, keep_symlinks, include_pat, exclude_pat, maxdepth, include_empty
+        name,
+        sources,
+        keep_symlinks=keep_symlinks,
+        include_pat=include_pat,
+        exclude_pat=exclude_pat,
+        maxdepth=maxdepth,
+        include_empty=include_empty,
+        merge=merge,
     )
 
-    for srelpath, ltarget in mng_symlinks:
+    for dirname in mng_dirs:
+        manage_directory(dirname)
+    for dest, src in mng_files:
+        manage_file(dest, src, replace)
+    # On Windows, we need symlink targets to exist, hence
+    # symlinks should be managed last.
+    for sdest, ltarget in mng_symlinks:
         _ret = symlink(
-            os.path.join(name, srelpath),
+            sdest,
             ltarget,
             makedirs=True,
             force=force_symlinks,
@@ -4923,11 +5056,7 @@ def recurse(
         )
         if not _ret:
             continue
-        merge_ret(os.path.join(name, srelpath), _ret)
-    for dirname in mng_dirs:
-        manage_directory(dirname)
-    for dest, src in mng_files:
-        manage_file(dest, src, replace)
+        merge_ret(sdest, _ret)
 
     if clean:
         # TODO: Use directory(clean=True) instead

--- a/salt/states/file.py
+++ b/salt/states/file.py
@@ -690,7 +690,15 @@ def _gen_recurse_managed_files(
         set(managed_files.items()),
         managed_directories,
         {
-            (link_path, link_target.replace(posixpath.sep, os.path.sep))
+            link_path: (
+                link_target.replace(posixpath.sep, os.path.sep),
+                os.path.normpath(
+                    os.path.join(
+                        os.path.dirname(link_path),
+                        link_target.replace(posixpath.sep, os.path.sep),
+                    )
+                ),
+            )
             for link_path, link_target in managed_symlinks.items()
         },
         keep,
@@ -5044,19 +5052,25 @@ def recurse(
         manage_file(dest, src, replace)
     # On Windows, we need symlink targets to exist, hence
     # symlinks should be managed last.
-    for sdest, ltarget in mng_symlinks:
-        _ret = symlink(
-            sdest,
-            ltarget,
-            makedirs=True,
-            force=force_symlinks,
-            user=user,
-            group=group,
-            mode=sym_mode,
-        )
-        if not _ret:
-            continue
-        merge_ret(sdest, _ret)
+    # In case there are n-level symlinks, we need to order them
+    # such that each target exists before the symlink is created.
+    processed_symlinks = set()
+    while to_process := set(mng_symlinks) - processed_symlinks:
+        for sdest in to_process:
+            ltarget, abstarget = mng_symlinks[sdest]
+            if abstarget in mng_symlinks and abstarget not in processed_symlinks:
+                continue
+            _ret = symlink(
+                sdest,
+                ltarget,
+                makedirs=True,
+                force=force_symlinks,
+                user=user,
+                group=group,
+                mode=sym_mode,
+            )
+            merge_ret(sdest, _ret)
+            processed_symlinks.add(sdest)
 
     if clean:
         # TODO: Use directory(clean=True) instead

--- a/tests/pytests/functional/fileserver/test_roots.py
+++ b/tests/pytests/functional/fileserver/test_roots.py
@@ -1,3 +1,6 @@
+import os.path
+import posixpath
+
 import pytest
 
 import salt.fileserver.roots as roots
@@ -20,7 +23,7 @@ def test_symlink_list(state_tree):
         link = state_tree / "link"
         link.symlink_to(str(target))
         ret = roots.symlink_list({"saltenv": "base"})
-        assert ret == {"link": str(target)}
+        assert ret == {"link": str(target).replace(os.path.sep, posixpath.sep)}
 
 
 @pytest.mark.parametrize(

--- a/tests/pytests/functional/states/file/test_recurse.py
+++ b/tests/pytests/functional/states/file/test_recurse.py
@@ -1,4 +1,8 @@
+import contextlib
+import copy
+import shutil
 import stat
+from pathlib import Path
 
 import pytest
 
@@ -206,6 +210,388 @@ def test_recurse_clean_specific_env(file, tmp_path):
     assert scene_32_dst.is_file() is False
     assert scene_32_dst.is_dir()
     assert scene_32_dst.joinpath("scene").is_file() is True
+
+
+@contextlib.contextmanager
+def create_file_tree(base, tree):
+    """
+    Helper to quickly create trees of directories, files and symlinks
+    inside a base directory.
+    """
+    temp_files = []
+    symlinks = {}
+    empty_dirs = []
+
+    def _create_tmp_files(tree, prefix):
+        for name, val in tree.items():
+            if isinstance(val, dict):
+                if not val:
+                    empty_dirs.append(base / prefix / name)
+                    continue
+                _create_tmp_files(val, prefix + f"/{name}")
+                continue
+            if isinstance(val, tuple):
+                symlinks[base / prefix / name] = val[0]
+                continue
+            temp_files.append(pytest.helpers.temp_file(name, val, base / prefix))
+
+    for first_level_name, first_level_contents in tree.items():
+        _create_tmp_files(first_level_contents, first_level_name)
+
+    with contextlib.ExitStack() as stack:
+        for file in temp_files:
+            stack.enter_context(file)
+        try:
+            created_dirs = []
+            for empty_dir in empty_dirs:
+                for par in empty_dir.parents[::-1]:
+                    if not par.exists():
+                        created_dirs.append(par)
+                        break
+                empty_dir.mkdir(parents=True)
+            for symlink, target in symlinks.items():
+                for par in symlink.parents[::-1]:
+                    if not par.exists():
+                        created_dirs.append(par)
+                        break
+                symlink.parent.mkdir(parents=True, exist_ok=True)
+                is_dir = target.endswith("/")
+                symlink.symlink_to(Path(target.rstrip("/")), target_is_directory=is_dir)
+                if is_dir:
+                    assert symlink.is_symlink()
+                    assert symlink.is_dir()
+            yield
+        finally:
+            for symlink in symlinks:
+                symlink.unlink(missing_ok=True)
+            for created_dir in created_dirs + empty_dirs:
+                shutil.rmtree(created_dir, ignore_errors=True)
+
+
+@pytest.fixture
+def _recurse_merge_dirs(state_tree, state_tree_prod):
+    tree_base = {
+        "file_from_base": "base",
+        "overridden_file": "foo",
+        "common": {
+            "file_from_base": "base",
+            "overridden_file": "foo",
+        },
+        "base": {
+            "present": "",
+        },
+        "empty": {
+            "in_base": {},
+            "common": {},
+            "override": {},
+            "override_file": "",
+            "nested_in_base_only": {
+                "worked": {},
+                "nonempty": {
+                    "file": "",
+                },
+            },
+        },
+        "deeply": "not_rendered",
+        "file_overridden_by_dir": "file",
+        "dir_overridden_by_file": {
+            "file_in_dir": "",
+        },
+        "symlinked_dir": ("symlink_target_dir/",),
+        "symlinked_file": ("symlink_target_file",),
+        "symlink_target_dir": {
+            "file_in_symlink_target": "",
+        },
+        "symlink_target_file": "",
+        "symlink_dir_overridden_by_file": ("symlink_target_dir/",),
+        "symlink_dir_overridden_by_dir": ("symlink_target_dir/",),
+        "symlink_file_overridden_by_dir": ("symlink_target_file",),
+        "symlink_file_overridden_by_file": ("symlink_target_file",),
+        "dir_overridden_by_symlink_dir": {
+            "file_in_dir_overridden_by_symlink_dir": "",
+        },
+        "dir_overridden_by_symlink_file": {
+            "file_in_dir_overridden_by_symlink_file": "",
+        },
+        "file_overridden_by_symlink_dir": "",
+        "file_overridden_by_symlink_file": "",
+        "symlink_file_overridden_by_symlink_file": ("symlink_target_file",),
+        "symlink_file_overridden_by_symlink_dir": ("symlink_target_file",),
+        "symlink_dir_overridden_by_symlink_file": ("symlink_target_dir/",),
+        "symlink_dir_overridden_by_symlink_dir": ("symlink_target_dir/",),
+        "symlink_target_dir_overridden_by_file": ("starget_dir_overridden_by_file/",),
+        "starget_dir_overridden_by_file": {
+            "file_in_starget_dir_overridden_by_file": "",
+        },
+        "symlink_target_dir_overridden_by_dir": ("starget_dir_overridden_by_dir/",),
+        "starget_dir_overridden_by_dir": {
+            "file_in_starget_dir_overridden_by_dir": "",
+        },
+        "symlink_target_file_overridden_by_dir": ("starget_file_overridden_by_dir",),
+        "starget_file_overridden_by_dir": "",
+        "symlink_target_file_overridden_by_file": ("starget_file_overridden_by_file",),
+        "starget_file_overridden_by_file": "",
+    }
+    tree_base["nested"] = copy.deepcopy(tree_base)
+    tree_override = {
+        "file_from_override": "override",
+        "overridden_file": "quux",
+        "common": {
+            "file_from_override": "override",
+            "overridden_file": "quux",
+        },
+        "empty": {
+            "in_override": {},
+            "common": {},
+            "override_dir": "",
+            "override_file": {},
+            "nested_in_base_only": {},
+            "symlinked": {
+                "dir": {},
+            },
+        },
+        "empty_symlinked_dir": ("empty/symlinked/",),
+        "empty_symlinked_dir_2": ("empty_symlinked_dir/",),
+        "deeply": {
+            "nested": {
+                "path": {
+                    "overrides_file": "success",
+                },
+            },
+        },
+        "override": {
+            "present": "",
+        },
+        "file_overridden_by_dir": {
+            "file_in_overriding_dir": "success",
+        },
+        "dir_overridden_by_file": "success",
+        "symlink_dir_overridden_by_dir": {
+            "file_in_dir_overriding_symlink_dir": "success",
+        },
+        "symlink_dir_overridden_by_file": "success",
+        "symlink_file_overridden_by_file": "success",
+        "symlink_file_overridden_by_dir": {
+            "file_in_dir_overriding_symlink_file": "success",
+        },
+        "other_symlink_target_dir": {
+            "other_file_in_symlink_target": "success",
+        },
+        "other_symlink_target_file": "success",
+        "dir_overridden_by_symlink_dir": ("other_symlink_target_dir/",),
+        "dir_overridden_by_symlink_file": ("other_symlink_target_file",),
+        "file_overridden_by_symlink_dir": ("other_symlink_target_dir/",),
+        "file_overridden_by_symlink_file": ("other_symlink_target_file",),
+        "symlink_file_overridden_by_symlink_file": ("other_symlink_target_file",),
+        "symlink_file_overridden_by_symlink_dir": ("other_symlink_target_dir/",),
+        "symlink_dir_overridden_by_symlink_file": ("other_symlink_target_file",),
+        "symlink_dir_overridden_by_symlink_dir": ("other_symlink_target_dir/",),
+        "starget_dir_overridden_by_file": "success",
+        "starget_dir_overridden_by_dir": {
+            "other_file_in_starget_dir_overridden_by_dir": "",
+        },
+        "starget_file_overridden_by_dir": {
+            "file_in_starget_file_overridden_by_dir": "success",
+        },
+        "starget_file_overridden_by_file": "success",
+        "symlink_to_parent": {
+            "dir": ("../other_symlink_target_dir/",),
+            "file": ("../file_from_override",),
+        },
+    }
+    tree_override["nested"] = copy.deepcopy(tree_override)
+
+    with create_file_tree(state_tree, {"base": tree_base}):
+        with create_file_tree(state_tree_prod, {"override": tree_override}):
+            yield
+
+
+@pytest.mark.usefixtures("_recurse_merge_dirs")
+@pytest.mark.parametrize("keep_symlinks", (False, True))
+@pytest.mark.parametrize("include_empty", (False, True))
+def test_recurse_merge(file, tmp_path, keep_symlinks, include_empty):
+    tgt = tmp_path / "target"
+    dirs = ("override", "base")
+    sources = [
+        f"salt://{src}" + ("?saltenv=prod" if src == "override" else "") for src in dirs
+    ]
+    ret = file.recurse(
+        name=str(tgt),
+        source=sources,
+        merge=True,
+        keep_symlinks=keep_symlinks,
+        include_empty=include_empty,
+    )
+    assert ret.result is True
+    assert tgt.is_dir()
+
+    # Run the same assertions for both the root path and a nested dir
+    for base in (tgt, tgt / "nested"):
+        # Ensure unique dirs are all merged
+        for tgt_dir in dirs:
+            assert (base / tgt_dir / "present").exists()
+        for shared_dir in (base, base / "common"):
+            # Ensure unique files in shared dir are all merged
+            for suffix in dirs:
+                assert (shared_dir / f"file_from_{suffix}").is_file()
+                assert (shared_dir / f"file_from_{suffix}").read_text() == suffix
+            # Ensure shared files are overridden by first source
+            assert (shared_dir / "overridden_file").is_file()
+            assert (shared_dir / "overridden_file").read_text() == "quux"
+        # Ensure dir <-> file overrides work
+        for path in (
+            ("file_overridden_by_dir", "file_in_overriding_dir"),
+            ("dir_overridden_by_file",),
+        ):
+            assert base.joinpath(*path).is_file()
+            assert base.joinpath(*path).read_text() == "success"
+        assert (base / "deeply" / "nested" / "path" / "overrides_file").exists()
+
+        # Sanity check include_empty
+        for empty_dir in ("in_base", "in_override", "common"):
+            assert ((base / "empty" / empty_dir).is_dir()) is include_empty
+        # Ensure empty dirs can be overridden by files
+        assert (base / "empty" / "override_dir").is_file()
+        # Ensure empty dirs can override files when include_empty is set
+        assert ((base / "empty" / "override_file").is_dir()) is include_empty
+        # Ensure empty dirs in overrides don't block nested dirs/files
+        assert (
+            (base / "empty" / "nested_in_base_only" / "worked").is_dir()
+        ) is include_empty
+        assert (base / "empty" / "nested_in_base_only" / "nonempty" / "file").is_file()
+
+        # Sanity check symlinks
+        assert ((base / "symlinked_dir").is_symlink()) is keep_symlinks
+        assert (base / "symlinked_dir").is_dir()
+        assert (base / "symlink_target_dir").is_dir()
+        for empty_symlink in ("empty_symlinked_dir", "empty_symlinked_dir_2"):
+            assert ((base / empty_symlink).exists()) is include_empty
+            assert ((base / empty_symlink).is_dir()) is include_empty
+            assert ((base / empty_symlink).is_symlink()) is (
+                include_empty and keep_symlinks
+            )
+
+        # Ensure symlinks can be overridden
+        for symlink_overridden in (
+            "symlink_dir_overridden_by_file",
+            "symlink_dir_overridden_by_dir",
+            "symlink_file_overridden_by_file",
+            "symlink_file_overridden_by_dir",
+        ):
+            assert not (base / symlink_overridden).is_symlink()
+            if symlink_overridden.endswith("file"):
+                assert (base / symlink_overridden).read_text() == "success"
+            else:
+                assert (
+                    base
+                    / symlink_overridden
+                    / (
+                        "file_in_dir_overriding_"
+                        + symlink_overridden.split("_overridden", maxsplit=1)[0]
+                    )
+                ).read_text() == "success"
+
+        # Ensure symlinks can override files/directories and other symlinks
+        for symlink in (
+            "dir_overridden_by_symlink_dir",
+            "dir_overridden_by_symlink_file",
+            "file_overridden_by_symlink_dir",
+            "file_overridden_by_symlink_file",
+            "symlink_dir_overridden_by_symlink_dir",
+            "symlink_dir_overridden_by_symlink_file",
+            "symlink_file_overridden_by_symlink_dir",
+            "symlink_file_overridden_by_symlink_file",
+        ):
+            assert ((base / symlink).is_symlink()) is keep_symlinks
+            if symlink.endswith("file"):
+                assert (base / symlink).read_text() == "success"
+            else:
+                assert (
+                    base / symlink / "other_file_in_symlink_target"
+                ).read_text() == "success"
+
+        # Ensure symlinked dirs are merged when not preserving symlinks
+        assert (
+            base / "symlink_dir_overridden_by_dir" / "file_in_symlink_target"
+        ).exists() is not keep_symlinks
+        assert (
+            base / "symlink_dir_overridden_by_symlink_dir" / "file_in_symlink_target"
+        ).exists() is not keep_symlinks
+        assert (
+            base
+            / "dir_overridden_by_symlink_dir"
+            / "file_in_dir_overridden_by_symlink_dir"
+        ).exists() is not keep_symlinks
+
+        # Ensure overridden symlink targets don't cause surprises
+        assert (base / "symlink_target_file_overridden_by_file").exists()
+        assert (
+            (base / "symlink_target_file_overridden_by_file").is_symlink()
+        ) is keep_symlinks
+        assert (
+            (base / "symlink_target_file_overridden_by_file").read_text() == "success"
+        ) is keep_symlinks
+        assert (
+            (base / "symlink_target_dir_overridden_by_dir").is_symlink()
+        ) is keep_symlinks
+        assert (base / "symlink_target_dir_overridden_by_dir").is_dir()
+
+        assert not (base / "symlink_target_dir_overridden_by_file").is_symlink()
+        assert (
+            (base / "symlink_target_dir_overridden_by_file").exists()
+        ) is not keep_symlinks
+        assert (
+            (base / "symlink_target_dir_overridden_by_file").is_dir()
+        ) is not keep_symlinks
+
+        assert not (base / "symlink_target_file_overridden_by_dir").is_symlink()
+        assert (
+            (base / "symlink_target_file_overridden_by_dir").exists()
+        ) is not keep_symlinks
+        assert (
+            (base / "symlink_target_file_overridden_by_dir").is_file()
+        ) is not keep_symlinks
+
+        # Ensure symlinks to parent paths works
+        assert (base / "symlink_to_parent/dir").exists()
+        assert (base / "symlink_to_parent/dir").is_symlink() is keep_symlinks
+        assert (base / "symlink_to_parent/dir").is_dir()
+        assert (base / "symlink_to_parent/dir/other_file_in_symlink_target").exists()
+        assert (base / "symlink_to_parent/file").exists()
+        assert (base / "symlink_to_parent/file").is_symlink() is keep_symlinks
+        assert (base / "symlink_to_parent/file").is_file()
+
+
+@pytest.mark.parametrize("include_empty", (False, True))
+def test_recurse_merge_keep_symlinks_broken_symlinks_are_kept(
+    file, state_tree, tmp_path, include_empty
+):
+    """
+    When ``keep_symlinks`` is set, broken symlinks should be kept.
+    They cause an exception when keep_symlinks is false
+    (and the fileserver follows them, ``fileserver_followsymlinks``).
+
+    This is not specific to ``merge=True``, but should be tested regardless.
+    """
+    tree = {
+        "base": {"foo": {"random_file": "", "broken_link": ("../nonexistent",)}},
+        "override": {"other_broken_link": ("42",)},
+    }
+    with create_file_tree(state_tree, tree):
+        tgt = tmp_path / "target"
+        dirs = ("override", "base")
+        sources = [f"salt://{src}" for src in dirs]
+        ret = file.recurse(
+            name=str(tgt),
+            source=sources,
+            merge=True,
+            keep_symlinks=True,
+            include_empty=include_empty,
+        )
+        assert ret.result is True
+        assert (tgt / "foo" / "broken_link").is_symlink()
+        assert (tgt / "foo" / "random_file").is_file()
+        assert (tgt / "other_broken_link").is_symlink()
 
 
 @pytest.mark.skip_on_windows(reason="'dir_mode' is not supported on Windows")

--- a/tests/pytests/functional/states/file/test_symlink.py
+++ b/tests/pytests/functional/states/file/test_symlink.py
@@ -40,3 +40,37 @@ def test_symlink_test(file, tmp_path):
     ret = file.symlink(str(symlink), target=str(target), test=True)
     assert ret.result is None
     assert symlink.exists() is False
+
+
+@pytest.mark.parametrize("typ", ("dir", "file"))
+@pytest.mark.parametrize("relation", ("parent", "sibling", "child"))
+def test_symlink_relative(file, tmp_path, typ, relation):
+    """
+    Ensure symlinks with relative targets work as expected.
+    This is especiallly important on windows, where symlinks
+    don't dynamically switch types between file and directory.
+    """
+    if relation == "parent":
+        symlink = tmp_path / "subdirectory" / "symlink"
+        target_spec = "../target"
+    elif relation == "sibling":
+        symlink = tmp_path / "symlink"
+        target_spec = "target"
+    else:
+        symlink = tmp_path / "symlink"
+        target_spec = "subdirectory/target"
+    target = (symlink.parent / target_spec).resolve()
+    if typ == "dir":
+        target.mkdir(parents=True)
+        check_file = symlink / "file_in_target"
+        (target / "file_in_target").write_text("resolved")
+    else:
+        target.parent.mkdir(parents=True, exist_ok=True)
+        target.write_text("resolved")
+        check_file = symlink
+    res = file.symlink(str(symlink), target=target_spec, makedirs=True)
+    assert res.result is True
+    assert symlink.exists()
+    assert symlink.is_symlink()
+    assert symlink.is_dir() is (typ == "dir")
+    assert check_file.read_text() == "resolved"

--- a/tests/pytests/unit/states/file/test_filestate.py
+++ b/tests/pytests/unit/states/file/test_filestate.py
@@ -258,22 +258,17 @@ def test_recurse():
             assert filestate.recurse(name, source) == ret
 
         with patch.object(os.path, "isabs", mock_t):
-            comt = "Invalid source '1' (must be a salt:// URI)"
+            comt = "Recurse failed: Invalid source '1' (must be a salt:// URI)"
             ret.update({"comment": comt})
             assert filestate.recurse(name, 1) == ret
 
-            comt = "Invalid source '//code/flask' (must be a salt:// URI)"
+            comt = (
+                "Recurse failed: Invalid source '//code/flask' (must be a salt:// URI)"
+            )
             ret.update({"comment": comt})
             assert filestate.recurse(name, "//code/flask") == ret
 
-            comt = "Recurse failed: "
-            ret.update({"comment": comt})
-            assert filestate.recurse(name, source) == ret
-
-            comt = (
-                "The directory 'code/flask' does not exist"
-                " on the salt fileserver in saltenv 'base'"
-            )
+            comt = "Recurse failed: none of the specified sources were found"
             ret.update({"comment": comt})
             assert filestate.recurse(name, source) == ret
 

--- a/tests/pytests/unit/states/file/test_recurse.py
+++ b/tests/pytests/unit/states/file/test_recurse.py
@@ -17,7 +17,8 @@ def configure_loader_modules():
 
 def test__gen_recurse_managed_files():
     """
-    Test _gen_recurse_managed_files to make sure it puts symlinks at the end of the list of files.
+    Test _gen_recurse_managed_files to make sure it does not include
+    symlinks in the list of files that is passed to file.managed.
     """
     target_dir = pathlib.Path(f"{os.sep}some{os.sep}path{os.sep}target")
     cp_list_master = MagicMock(
@@ -28,6 +29,12 @@ def test__gen_recurse_managed_files():
             "target/notasymlink",
         ],
     )
+    cp_list_master_dirs = MagicMock(
+        return_value=[
+            "target",
+            "target/not_a_symlink",
+        ],
+    )
     cp_list_master_symlinks = MagicMock(
         return_value={
             "target/symlink": f"{target_dir}{os.sep}not_a_symlink{os.sep}symlink"
@@ -35,14 +42,22 @@ def test__gen_recurse_managed_files():
     )
     patch_salt = {
         "cp.list_master": cp_list_master,
+        "cp.list_master_dirs": cp_list_master_dirs,
         "cp.list_master_symlinks": cp_list_master_symlinks,
     }
     with patch.dict(filestate.__salt__, patch_salt):
         files, dirs, links, keep = filestate._gen_recurse_managed_files(
-            name=str(target_dir), source=f"salt://{target_dir.name}", keep_symlinks=True
+            name=str(target_dir),
+            sources=[f"salt://{target_dir.name}"],
+            keep_symlinks=True,
         )
-    expected = (
+    unexpected = (
         f"{os.sep}some{os.sep}path{os.sep}target{os.sep}symlink",
         "salt://target/symlink?saltenv=base",
     )
-    assert files[-1] == expected
+    assert unexpected not in files
+    expected = (
+        f"{os.sep}some{os.sep}path{os.sep}target{os.sep}symlink",
+        f"{os.sep}some{os.sep}path{os.sep}target{os.sep}not_a_symlink{os.sep}symlink",
+    )
+    assert expected in links

--- a/tests/pytests/unit/states/file/test_recurse.py
+++ b/tests/pytests/unit/states/file/test_recurse.py
@@ -56,8 +56,10 @@ def test__gen_recurse_managed_files():
         "salt://target/symlink?saltenv=base",
     )
     assert unexpected not in files
+    expected_dest = f"{os.sep}some{os.sep}path{os.sep}target{os.sep}symlink"
     expected = (
-        f"{os.sep}some{os.sep}path{os.sep}target{os.sep}symlink",
+        f"{os.sep}some{os.sep}path{os.sep}target{os.sep}not_a_symlink{os.sep}symlink",
         f"{os.sep}some{os.sep}path{os.sep}target{os.sep}not_a_symlink{os.sep}symlink",
     )
-    assert expected in links
+    assert expected_dest in links
+    assert links[expected_dest] == expected

--- a/tests/pytests/unit/states/test_file.py
+++ b/tests/pytests/unit/states/test_file.py
@@ -81,11 +81,10 @@ def test_file_recurse_directory_test():
         ret = file.recurse("/tmp/test", "salt://does_not_exist", saltenv="base")
         assert ret == {
             "changes": {},
-            "comment": "The directory 'does_not_exist' does not exist on the salt fileserver in saltenv 'base'",
+            "comment": "Recurse failed: none of the specified sources were found",
             "name": "/tmp/test",
             "result": False,
         }
         salt_dunder["cp.list_master_dirs"].assert_called_once_with(
             saltenv="base",
-            prefix="does_not_exist/",
         )

